[EDNA-14] Add OCI image packages

mkaito opened the pull request:
## Description

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

OCI images for backend and frontend containers. Built but not (yet) pushed in CI.

Build locally with `nix build .#docker-backend`.

Import [as documented](https://docs.docker.com/engine/reference/commandline/import/#import-from-a-local-file), then run the container. It expects environment variables `$API_PORT` and `$SERVE_DOCS` set.

Config layering is tentative until I figure out how exactly the backend needs to be configured.

Related PR: https://github.com/serokell/gemini-infra/pull/4

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

https://issues.serokell.io/issue/EDNA-14

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

Commit:
[EDNA-14] Add OCI image packages

gromakovsky commented on flake.nix:
```diff
@@ -20,38 +21,46 @@
+          nix-npm-buildpackage.overlay
+          haskell-nix.overlay
+          serokell-nix.overlay
+        ];
```
Oh nice, looks more understandable this way.

gromakovsky commented on docker.nix:
```diff
@@ -0,0 +1,99 @@
+    /bin/envsubst '$API_PORT $SERVE_DOCS' \
+      < ${backend-config-template} >| /config.yaml
+
+    exec /bin/edna-server -c -c /config.yaml "$@"
```
Looks like redundant duplicated `-c`.

mkaito replied:
I have no idea how it is run or configured yet :stuck_out_tongue: I need to have a look at that.

gromakovsky commented on docker.nix:
```diff
@@ -0,0 +1,99 @@
+in
+{
+  backend-image = buildLayeredImage {
+    name = "ghcr.io/serokell/edna-backend";
```
Is it the place where the image will be uploaded?

mkaito replied:
Well, yes. But also, no. The image name must match the registry URI where it will be available. This name does not necessarily imply that it is being pushed there now. It's just a name.

balsoft replied:
I don't think this name _must_ match the registry URI.

balsoft commented on docker.nix:
```diff
@@ -0,0 +1,99 @@
+    echo "nobody:x::" > $out/etc/gshadow
+  '';
+
+  backend-config-template = writeText "backend-config-env.yaml.template" (toJSON {
```
BTW you can import the `config.yaml` that's in the repo here and merge it with those overrides. See https://gitlab.com/tezosagora/agora/-/blob/master/nix/modules/services/backend.nix#L61 for example

mkaito replied:
Oh, nice!

balsoft commented on docker.nix:
```diff
@@ -0,0 +1,99 @@
+  httpPort = 80;
+  httpsPort = 443;
+
+  # Some system files necessary for nginx to function normally
```
nginx?

mkaito commented:
The following environment variables are available at runtime in the backend container:

          POSTGRES_HOST 
          POSTGRES_DB
          POSTGRES_USER
          POSTGRES_PASSWORD

I can interpolate them into a config file, or you can consume them from the environment once you get around to postgres usage.

The staging server's canonical name is `castor.gemini.serokell.team`. It also has a CNAME `staging.edna.serokell.team` so you don't have to remember it.

I'm still fiddling with a few runtime things, but feel free to review.

gromakovsky commented on .buildkite/pipeline.yml:
```diff
@@ -11,12 +11,18 @@ steps:
       - nix-build -A packages.x86_64-linux.frontend
+  - label: frontend image
+    commands:
+      - nix-build -A packages.x86_64-linux.docker-frontend
```
I propose to add built images into CI artifacts to be able to quickly check what was built on CI.

mkaito replied:
Sure, why not.

gromakovsky commented:
![image](https://user-images.githubusercontent.com/2556409/111026187-17d79880-83fa-11eb-9f09-2ad6036f7a61.png)
I am a Docker newbie :grimacing: So I imported 2 `tar.gz` files (not sure if correctly btw) and got 2 hashes. What should I do next to run them?

balsoft commented:
You shouldn't `docker import` those images, but rather `docker load -i` them. Then you'll be able to `docker run ghcr.io/serokell/edna-frontend` and `docker run ghcr.io/serokell/edna-backend`

mkaito commented:
Yeah, sorry. I'm not sure if the command semantics changed or my memory got corrrupted since the last time I worked with this kind of thing. @balsoft is correct.

zhenyavinogradov commented on frontend/default.nix:
```diff
@@ -1,15 +1,21 @@
+
+buildYarnPackage {
+  SASS_BINARY_PATH = fetchurl {
+    url = "https://github.com/sass/node-sass/releases/download/v4.12.0/linux-x64-64_binding.node";
```
In the yarn lockfile it's `v4.14.1` instead of `v4.12.0`, I think it should be the same version here

mkaito replied:
Updated

gromakovsky commented:
@mkaito we have merged 2 PRs related to DB integration, so now 2 your PRs (this one and in another repo) need to be updated a bit. Specifically:
* `edna-server` now connects to a PostgreSQL DB. This is how it's configured:
```
db:
  # Syntax: https://www.postgresql.org/docs/9.5/libpq-connect.html#LIBPQ-CONNSTRING
  conn-string: "host=localhost dbname=ednadb user=ednauser password=ednapass"
  max-connection: 200
```
`max-connection: 200` is probably ok.
* Also there is an optional configuration setting to init the DB:
```
  # Optional value used if you want to init DB on start
  initiation:
    # Two options of DB initialization from the app:
    # * "enable" will run init script on the app start,
    # * "enable-with-drop" will run drop script before init
    mode: "enable"
    init-script: "./sql/init.sql"
```

For staging, I think we can create DB from scratch on each push to master, so that we don't need to care about schema changes and migrations for now. So I think in the staging config `mode` should be `"enable-with-drop"` and you should make sure that `init-script` points to https://github.com/serokell/edna/blob/1a98cf6a37da384f7c23180c9db132e82e030803/backend/sql/init.sql

Is it ok, can you do it?

zhenyavinogradov commented on flake.nix:
```diff
@@ -20,59 +24,109 @@
+    deploy.nodes.castor =
+      let
+        # pkgs.symlinkJoin et al can only handle folders, not files; so we roll
+        # our own, but skip the `dirname` call.
```
There is `linkFarm`, which can handle files

mkaito replied:
I tried `linkFarm` and it complained about it being a file.

mkaito replied:
Nevermind, seems I'm just dumb.

gromakovsky commented on docker.nix:
```diff
@@ -0,0 +1,75 @@
+
+    config = {
+      Entrypoint = "/bin/entrypoint.sh";
+      Cmd = [ "-c" "/config.yaml" ];
```
Where does it come from? AFAIU this `/config.yaml` is a path inside the OCI image that we are creating here. In which place do we create this file in the image?

mkaito replied:
You're expected to provide it externally via mount or volume.

To provide more context, `Cmd` are given as arguments to `Entrypoint`. When you `docker run foo:latest banana`, `banana` overrides the `Cmd` portion. So it's just a default.

gromakovsky replied:
So AFAIU the relevant part of `edna.nix` (in another repo) is this:
```
        volumes = [
          "${configFile}:/config.yaml:ro"
        ];

        cmd = [
          "-c" "/config.yaml"
        ];
```

mkaito replied:
Yes. I've decided to remove the default Cmd from this repo, since it depends on where the user mounts the config file anyway.

gromakovsky committed:
Rename some configuration options

Problem: two configuration options are called a bit weirdly.
Solution: rename them.

gromakovsky reviewed (approved):
Looks plausible.

gromakovsky committed:
Add a simple log message to edna-server

Problem: during debug sometimes one may wonder whether edna-server
actually started. And it may be non-trivial to figure it out.
Solution: print a simple log message when edna-server is started.

gromakovsky commented on docker.nix:
```diff
@@ -0,0 +1,84 @@
+    # we use runCommand and copy the file instead.
+    init-sql = runCommand "edna-init-sql" {} ''
+      mkdir -p $out
+      cp ${./backend/sql/init.sql} $out/init.sql
```
Minor suggestion: maybe let's put it into `$out/sql/init.sql` so that `dev-config.yaml` works? In `dev-config.yaml` we have `init-script: "./sql/init.sql"`.
